crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1935 @@
+ * 56 bits apart, but are greater than 56 bits in length, the most significant
+ * bits of each limb overlap with the least significant bits of the next
+ *
+ * This representation is considered to be 'redundant' in the sense that
+ * intermediate values can each contain more than a 56-bit value in each limb.
+ * Reduction causes all but the final limb to be reduced to contain a value less
+ * than 2^56, with the final value represented allowed to be larger than 2^384,
+ * inasmuch as we can be sure that arithmetic overflow remains impossible. The
+ * reduced value must of course be congruent to the unreduced value.
+ *
+ * A field element with 64-bit limbs is an 'felem'. One with 128-bit limbs is a
+ * 'widefelem', featuring enough bits to store the result of a multiplication
+ * and even some further arithmetic without need for immediate reduction.
+ */
+
+#define NLIMBS 7
+
+typedef uint64_t limb;
+typedef uint128_t widelimb;
+typedef limb limb_aX __attribute((__aligned__(1)));
+typedef limb felem[NLIMBS];
+typedef widelimb widefelem[2*NLIMBS-1];
+
+static const limb bottom56bits = 0xffffffffffffff;
+
+/* Helper functions (de)serialising reduced field elements in little endian */
+static void bin48_to_felem(felem out, const u8 in[48])
+{
+    memset(out, 0, 56);
+    out[0] = (*((limb *) & in[0])) & bottom56bits;
+    out[1] = (*((limb_aX *) & in[7])) & bottom56bits;
+    out[2] = (*((limb_aX *) & in[14])) & bottom56bits;
+    out[3] = (*((limb_aX *) & in[21])) & bottom56bits;
+    out[4] = (*((limb_aX *) & in[28])) & bottom56bits;
+    out[5] = (*((limb_aX *) & in[35])) & bottom56bits;
+    memmove(&out[6], &in[42], 6);
+}
+
+static void felem_to_bin48(u8 out[48], const felem in)
+{
+    memset(out, 0, 48);
+    (*((limb *) & out[0]))     |= (in[0] & bottom56bits);
+    (*((limb_aX *) & out[7]))  |= (in[1] & bottom56bits);
+    (*((limb_aX *) & out[14])) |= (in[2] & bottom56bits);
+    (*((limb_aX *) & out[21])) |= (in[3] & bottom56bits);
+    (*((limb_aX *) & out[28])) |= (in[4] & bottom56bits);
+    (*((limb_aX *) & out[35])) |= (in[5] & bottom56bits);
+    memmove(&out[42], &in[6], 6);
+}
+
+/* BN_to_felem converts an OpenSSL BIGNUM into an felem */
+static int BN_to_felem(felem out, const BIGNUM *bn)
+{
+    felem_bytearray b_out;
+    int num_bytes;
+
+    if (BN_is_negative(bn)) {
+        ERR_raise(ERR_LIB_EC, EC_R_BIGNUM_OUT_OF_RANGE);
+        return 0;
+    }
+    num_bytes = BN_bn2lebinpad(bn, b_out, sizeof(b_out));
+    if (num_bytes < 0) {
+        ERR_raise(ERR_LIB_EC, EC_R_BIGNUM_OUT_OF_RANGE);
+        return 0;
+    }
+    bin48_to_felem(out, b_out);
+    return 1;
+}
+
+/* felem_to_BN converts an felem into an OpenSSL BIGNUM */
+static BIGNUM *felem_to_BN(BIGNUM *out, const felem in)
+{
+    felem_bytearray b_out;
+    felem_to_bin48(b_out, in);
+    return BN_lebin2bn(b_out, sizeof(b_out), out);
+}
+
+/*-
+ * Field operations
+ * ----------------
+ */
+
+static void felem_one(felem out)
+{
+    unsigned i;
```

Comment:
Our coding style uses `unsigned int` here and elsewhere.

Also a blank line separating declarations and code, again here and elsewhere.